File: packages/typescript/src/types/events.ts
import type { Market, MarketStatus } from './markets.js';
import type { PaginationParams, SortField } from './common.js';

/**
 * Settlement source information for an event.
 */
export interface SettlementSource {
  /** Name of the settlement source */
  name: string;
  /** URL of the settlement source */
  url: string;
}

/**
 * Event data returned from the API.
 */
export interface Event {
  /** Unique ticker identifier for the event */
  ticker: string;
  /** Title of the event */
  title: string;
  /** Subtitle of the event */
  subtitle: string;
  /** Series ticker this event belongs to */
  seriesTicker: string;
  /** Competition name (optional) */
  competition?: string | null;
  /** Competition scope (optional) */
  competitionScope?: string | null;
  /** URL to the event image (optional) */
  imageUrl?: string | null;
  /** Total liquidity in the event (optional) */
  liquidity?: number | null;
  /** Markets within this event (only included when withNestedMarkets is true) */
  markets?: Market[] | null;
  /** Open interest in the event (optional) */
  openInterest?: number | null;
  /** Settlement sources for the event (optional) */
  settlementSources?: SettlementSource[] | null;
  /** Strike date timestamp in seconds (optional) */
  strikeDate?: number | null;
  /** Strike period description (optional) */
  strikePeriod?: string | null;
  /** Total volume traded (optional) */
  volume?: number | null;
  /** 24-hour trading volume (optional) */
  volume24h?: number | null;
}

/**
 * Parameters for fetching events.
 */
export interface EventsParams extends PaginationParams {
  /** Filter events by market status */
  status?: MarketStatus;
  /** 
   * Filter by series tickers (comma-separated list, max 25).
   * Example: "KXBTC,KXETH"
   */
  seriesTickers?: string;
  /** Include nested markets in response */
  withNestedMarkets?: boolean;
  /** Filter events that are initialized (have a corresponding market ledger) */
  isInitialized?: boolean;
  /** Sort field for results */
  sort?: SortField;
}

/**
 * Response from the events endpoint.
 */
export interface EventsResponse {
  /** Pagination cursor for next page (number of events to skip) */
  cursor?: number | null;
  /** Array of events */
  events: Event[];
}

/**
 * Parameters for fetching forecast percentile history.
 */
export interface ForecastHistoryParams {
  /** 
   * Comma-separated list of percentile values (0-10000, max 10 values).
   * Example: "2500,5000,7500" for 25th, 50th, 75th percentiles
   */
  percentiles: string;
  /** Start timestamp for the range (Unix timestamp in seconds) */
  startTs: number;
  /** End timestamp for the range (Unix timestamp in seconds) */
  endTs: number;
  /** Period interval in minutes (0, 1, 60, or 1440) */
  periodInterval: number;
}

/**
 * Forecast history point data.
 */
export interface ForecastHistoryPoint {
  /** Timestamp of the data point */
  timestamp: number;
  /** YES price at this point */
  yesPrice: number;
  /** NO price at this point */
  noPrice: number;
  /** Percentile value (optional) */
  percentile?: number;
}

/**
 * Forecast history response.
 * Note: This endpoint relays the response directly from the Kalshi API.
 */
export interface ForecastHistory {
  /** Event ticker */
  eventTicker: string;
  /** Array of historical forecast data points */
  history: ForecastHistoryPoint[];
}

File: packages/python/src/dflow/types/events.py
"""Event types for DFlow SDK."""

from pydantic import BaseModel, Field

from .markets import Market, MarketStatus


class SettlementSource(BaseModel):
    """Settlement source information for an event."""

    name: str
    url: str


class Event(BaseModel):
    """Prediction event data."""

    ticker: str
    title: str
    subtitle: str
    series_ticker: str = Field(alias="seriesTicker")

    # Optional fields from API
    competition: str | None = None
    competition_scope: str | None = Field(default=None, alias="competitionScope")
    image_url: str | None = Field(default=None, alias="imageUrl")
    liquidity: float | None = None
    markets: list[Market] | None = None
    open_interest: float | None = Field(default=None, alias="openInterest")
    settlement_sources: list[SettlementSource] | None = Field(
        default=None, alias="settlementSources"
    )
    strike_date: int | None = Field(default=None, alias="strikeDate")
    strike_period: str | None = Field(default=None, alias="strikePeriod")
    mutually_exclusive: bool | None = Field(default=None, alias="mutuallyExclusive")
    volume: float | None = None
    volume_24h: float | None = Field(default=None, alias="volume24h")

    model_config = {"populate_by_name": True}


class EventsParams(BaseModel):
    """Parameters for fetching events."""

    # Filter events by market status
    status: MarketStatus | None = None
    # Filter by series tickers (comma-separated list, max 25)
    series_tickers: str | None = Field(default=None, alias="seriesTickers")
    # Include nested markets in response
    with_nested_markets: bool | None = Field(default=None, alias="withNestedMarkets")
    # Filter events that are initialized (have a corresponding market ledger)
    is_initialized: bool | None = Field(default=None, alias="isInitialized")
    # Sort field for results
    sort: str | None = None
    # Pagination cursor
    cursor: int | None = None
    # Limit
    limit: int | None = None

    model_config = {"populate_by_name": True}


class EventsResponse(BaseModel):
    """Response from events list endpoint."""

    cursor: int | None = None
    events: list[Event]


class ForecastHistoryParams(BaseModel):
    """Parameters for fetching forecast percentile history."""

    # Comma-separated list of percentile values (0-10000, max 10 values)
    percentiles: str
    # Start timestamp for the range (Unix timestamp in seconds)
    start_ts: int = Field(alias="startTs")
    # End timestamp for the range (Unix timestamp in seconds)
    end_ts: int = Field(alias="endTs")
    # Period interval in minutes (0, 1, 60, or 1440)
    period_interval: int = Field(alias="periodInterval")

    model_config = {"populate_by_name": True}


class ForecastHistoryPoint(BaseModel):
    """Single point in forecast history."""

    timestamp: int
    yes_price: float = Field(alias="yesPrice")
    no_price: float = Field(alias="noPrice")
    percentile: float | None = None

    model_config = {"populate_by_name": True}


class ForecastHistory(BaseModel):
    """Forecast history data for an event.

    Note: This endpoint relays the response directly from the Kalshi API.
    """

    event_ticker: str = Field(alias="eventTicker")
    history: list[ForecastHistoryPoint]

    model_config = {"populate_by_name": True}
